datafusion/core/src/execution/context.rs

```diff
@@ -564,8 +568,58 @@ impl SessionContext {
         }
 
         // create a query planner
-        let state = self.state.read().clone();
-        let query_planner = SqlToRel::new(&state);
+        let mut provider = {
```

Comment:
This is the real meat of the change, rather than making planning async, we instead determine what tables are needed by the query ahead of time and use those.